src/BatteryStateOfChargeService.cpp

```diff
@@ -35,29 +39,58 @@ namespace
 BatteryStateOfChargeService::BatteryStateOfChargeService(double initialStateOfChargePercent)
 : initialStateOfChargePercent_(initialStateOfChargePercent)
 {
+ AmpHoursUsed_=BATTERY_AMP_HOUR_CAPACITY*(initialStateOfChargePercent_/100);
 }
-
 BatteryStateOfChargeService::~BatteryStateOfChargeService()
 {
 }
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return (AmpHoursUsed_);
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
+    if(newCurrent_>=0)
     return false;
+    else
+    return true;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+  double timeLeft;
+  timeLeft=(AmpHoursUsed_/newCurrent_);
+  timeLeft=qAbs(timeLeft);
+  timeLeft=timeLeft/2.77778e-7;
+
+  QTime base(0,0,0);
+  QTime timeOutput= base.addMSecs(timeLeft);
+  return timeOutput;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+ //Setting Values
+
+    inCurrent_=newCurrent_;
+    initialTime_=currentTime_;
+
+    newCurrent_=batteryData.current;
+    currentTime_=batteryData.time;
+
+    if(firstRun_!=true){
+
+    int mschangeTime=abs(initialTime_.msecsTo(currentTime_));
+    double changeTime=mschangeTime*2.77778e-7;
```

Comment:
This value `2.77778e-7` is something that we usually call a "magic number", it's a number that does something but has no context as to what it actually does. This number should be abstracted into a constant like how I described in `LogFileReader.cpp`.